builder/test/cse.cc

```diff
@@ -0,0 +1,166 @@
+                                         << actual << "\ninstead of:\n"
+                                         << expected << "\n";
+}
+
+}  // namespace
+
+TEST(cse, no_op) {
+  node_context ctx = symbols;
+  expr e = ssa_block(ctx, {abs(x), t0 * t0});
+  // This is fine as-is.
+  test(ctx, e, e);
+}
+
+TEST(cse, simple) {
+  node_context ctx = symbols;
+  expr e = ((x * x + x) * (x * x + x)) + x * x;
+  e += e;
+  expr correct = ssa_block(ctx, {x * x,            // x*x
+                                    x + t0,        // x*x + x
+                                    t1 * t1 + t0,  // (x*x + x)*(x*x + x) + x*x
+                                    t2 + t2});
+  // Test a simple case.
+  test(ctx, e, correct);
+
+  // Check for idempotence (also checks a case with lets)
+  test(ctx, correct, correct);
+}
+
+TEST(cse, redundant_lets) {
+  node_context ctx = symbols;
+  expr e = ssa_block(ctx, {x * x, x * x, t0 / t1, t1 / t1, t2 % t3, (t4 + x * x) + x * x});
+  expr correct = ssa_block(ctx, {x * x, t0 / t0, (t1 % t1 + t0) + t0});
+  test(ctx, e, correct);
+}
+
+TEST(cse, nested_lets) {
+  node_context ctx = symbols;
+  expr e1 = ssa_block(ctx, {x * x,              // a = x*x
+                               t0 + x,          // b = a + x
+                               t1 * t1 * t0});  // c = b * b * a
+  expr e2 = ssa_block(ctx, {x * x,              // a again
+                               t0 - x,          // d = a - x
+                               t1 * t1 * t0});  // e = d * d * a
+  expr e = ssa_block(ctx, {e1 + x * x,          // f = c + a
+                              e1 + e2,          // g = c + e
+                              t0 + t0 * t1});   // h = f + f * g
+
+  expr correct = ssa_block(ctx, {x * x,                               // t0 = a = x*x
+                                    x + t0,                           // t1 = b = a + x     = t0 + x
+                                    t1 * t1 * t0,                     // t2 = c = b * b * a = t1 * t1 * t0
+                                    t2 + t0,                          // t3 = f = c + a     = t2 + t0
+                                    t0 - x,                           // t4 = d = a - x     = t0 - x
+                                    t3 + t3 * (t2 + t4 * t4 * t0)});  // h (with g substituted in)
+  test(ctx, e, correct);
+}
+
+TEST(cse, scales_ok) {
+  node_context ctx = symbols;
+  expr e = x;
+  // Test against pathological runtimes (don't bother checking correctness)
+  for (int i = 0; i < 100; i++) {
+    e = e * e + e + i;
+    e = e * e - e * i;
+  }
+  expr result = common_subexpression_elimination(e, ctx);
+}
+
+TEST(cse, select) {
+  node_context ctx = symbols;
+
+  expr buf = ctx.insert_unique("buf");
+  expr index = select(x * x + y * y > 0, x * x + y * y + 2, x * x + y * y + 10);
+  expr at_args[] = {index};
+  expr load = buffer_at(buf, at_args);
+
+  expr e = select(x * y > 10, x * y + 2, x * y + 3 + load);
```

Comment:
I don't understand the comment here -- why is the location a problem?

Reply:
Because I want to know (without reading carefully for a single character variable) where this expression is used. If you move it to immediately before its first use, it reduces the number of places it could be used.